### PR TITLE
[balsa] Run HttpProtocolIntegrationTests with BalsaParser as well.

### PR DESCRIPTION
**Pull request by bencebeky**
Set up HttpProtocolIntegrationTest and its derived test classes, as well as Http2BufferWatermarksTest (a class derived from HttpIntegrationTest) to run with BalsaParser in addition to http-parser if either the upstream or the downstream protocol is HTTP/1.1.  Disable BalsaParser for now for a few tests that are failing.

Also bump shard counts for the following two tests:

//test/integration:overload_integration_test
//test/extensions/filters/http/health_check:health_check_integration_test

This is because the increased number of test cases resulted in timeouts (against the five minute limit) of TSAN runs.

Tracking issue: https://github.com/envoyproxy/envoy/issues/21245

Signed-off-by: Bence Béky <bnc@google.com>

Commit Message: [balsa] Run HttpProtocolIntegrationTests with BalsaParser as well.
Additional Description:
Risk Level: low, test-only change
Testing: //test/integration:all
Docs Changes: n/a
Release Notes: n/a
Platform Specific Features: n/a

### CODE REVIEW

File: test/integration/http_protocol_integration.cc [hunk 1]
```diff
@@ -8,42 +8,40 @@ std::vector<HttpProtocolTestParams> HttpProtocolIntegrationTest::getProtocolTest
...
-#ifdef ENVOY_ENABLE_QUIC
-        ret.push_back(HttpProtocolTestParams{ip_version, downstream_protocol, upstream_protocol,
-                                             Http2Impl::Nghttp2, false});
-        if (downstream_protocol == Http::CodecType::HTTP2 ||
-            upstream_protocol == Http::CodecType::HTTP2) {
-          addHttp2TestParametersWithDeferredProcessing(ip_version, downstream_protocol,
-                                                       upstream_protocol);
-        }
-#else
+#ifndef ENVOY_ENABLE_QUIC
```
Comment (diannahu) [inline test/integration/http_protocol_integration.cc:14]: Nice, this looks cleaner!

File: test/extensions/filters/http/health_check/BUILD [hunk 2]
```diff
@@ -45,6 +45,7 @@ envoy_extension_cc_test(
         "health_check_integration_test.cc",
     ],
     extension_names = ["envoy.filters.http.health_check"],
+    shard_count = 4,
```
Comment (diannahu) [inline test/extensions/filters/http/health_check/BUILD:48]: no action, just curious here and in the other shard count bump: Did this help with the speed of the tests?

**Review by diannahu — APPROVED** [review]
Nice test enablement!

**Review by diannahu — APPROVED** [review]

**Review by KBaichoo — APPROVED** [review]
looks good, thank you!

**Pull request merged**